src/main/java/com/google/graphgeckos/dashboard/storage/DatastoreRepository.java

```diff
@@ -0,0 +1,156 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.graphgeckos.dashboard.storage;
+
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.Query;
+import com.google.cloud.datastore.StructuredQuery.OrderBy;
+import com.google.cloud.datastore.StructuredQuery.PropertyFilter;
+import com.google.cloud.datastore.TimestampValue;
+import com.google.cloud.Timestamp;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.List;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.SpringApplication;
+import org.springframework.cloud.gcp.data.datastore.core.DatastoreTemplate;
+import org.springframework.stereotype.Repository;
+
+/**
+ * A DataRepository implementation backed up by Google Datastore.
+ * Datastore. Each database entry is modeled by the {@link #BuildInfo BuildInfo} class.
+ * The relevant fields for the database are:
+ *    - Kind: "revision"
+ *    - Key: commit hash
+ *
+ * Useful links:
+ * - https://googleapis.dev/java/google-cloud-datastore/latest/index.html
+ * - https://googleapis.dev/java/spring-cloud-gcp/1.2.2.RELEASE/index.html
+ * - https://cloud.google.com/datastore/docs/concepts/
+ * - https://spring.io/projects/spring-cloud-gcp#overview
+ */
+@Repository
+public class DatastoreRepository implements DataRepository {
+  @Autowired
+  private DatastoreTemplate storage;
+
+  /**
+   * {@inheritDoc}
+   */
+  public boolean createRevisionEntry(ParsedGitData entryData) throws IllegalArgumentException {
+    if (entryData == null) {
+      throw new IllegalArgumentException("entryData cannot be null");
+    }
+
+    if (getRevisionEntry(entryData.getCommitHash()) == null) {
+      try {
+        storage.save(new BuildInfo(entryData));
+      } catch (Exception e) {
```

Comment:
This catch is very broad; I'd not catch it here but let it bubble up to the caller for now.
If an exception happens here, it's probably because of some connectivity or concurrence issue, so just returning `false` in that case potentially discards relevant exception information that may help with debugging. Alternatively, I'd suggest to log the exception so that we have a change to see and debug what happened.

Reply:
the .save(T instance) method uses the DatastoreReaderWriter::put method, which by the Java Reference states:
`put
Entity put(FullEntity<?> entity)
A Datastore put (a.k.a upsert) operation: inserts an entity if it does not exist, updates it otherwise. This method will automatically allocate an id if necessary.
Parameters:
entity - the entity to put
Returns:
an Entity with the same properties and a key that is either newly allocated or the same one if key is already complete
Throws:
DatastoreException - upon failure
IllegalArgumentException - if the given entity is missing a key`

So, basically in this case we can only theoretically get a DatastoreException or an IllegalArgumentException. Since we sanitize the inputs at every stage, the IllegalArgumentException would have been thrown earlier, which leaves only the DatastoreException. If I change `Exception` to `DatastoreException`, would it make sense to leave the rest of the code as-is?

Reply:
Great!
> If I change Exception to DatastoreException, would it make sense to leave the rest of the code as-is?

That would be OK. I'd suggest still logging the DatastoreException in the catch clause in that case so we have a chance to see the details about such problems at runtime.

Reply:
For now, the stack trace and exception + exception message are printed to stderr.